runtime/service.go

```diff
@@ -537,25 +542,62 @@ func dialVsock(ctx context.Context, contextID uint32, port uint32) (net.Conn, er
 	return nil, lastErr
 }
 
-func (s *service) startVM(ctx context.Context, request *taskAPI.CreateTaskRequest) (taskAPI.TaskService, error) {
-	/*
-		What needs to be done here:
-			- Start a firecracker agent with:
-				- The container rootfs as a block device
-					- rootfs will be the cwd
-				- Mount this device inside agent at a well known location
-				- With the agent running inside on a well-known port
-			- After agent startup, create a vsock client dialing to the
-				specified vsock port.
-				TODO: We will need some sort of vsock CID accounting mechanism
-				to know what CID to use, defaults to 3 for now
-			- Return this client or error
-	*/
+// findNextAvailableVsockCID finds first available vsock context ID.
+// It uses VHOST_VSOCK_SET_GUEST_CID ioctl which allows some CID ranges to be statically reserved in advance.
+// The ioctl fails with EADDRINUSE if cid is already taken and with EINVAL if the CID is invalid.
+// Taken from https://bugzilla.redhat.com/show_bug.cgi?id=1291851
+func findNextAvailableVsockCID(ctx context.Context) (uint32, error) {
+	const (
+		// Corresponds to VHOST_VSOCK_SET_GUEST_CID in vhost.h
+		ioctlVsockSetGuestCID = uintptr(0x4008AF60)
+		// 0, 1 and 2 are reserved CIDs, see http://man7.org/linux/man-pages/man7/vsock.7.html
+		startCID        = 3
+		maxCID          = math.MaxUint32
+		vsockDevicePath = "/dev/vhost-vsock"
+	)
+
+	file, err := os.OpenFile(vsockDevicePath, syscall.O_RDWR, 0666)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to open vsock device")
+	}
+
+	defer file.Close()
 
+	for contextID := startCID; contextID < maxCID; contextID++ {
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		default:
+			cid := contextID
+			_, _, err = sysCall(
```

Comment:
Why not just use `syscall.Syscall` here?

Reply:
To make it testable. See thread with Ben above.

Reply:
Got it, thanks.